engines/native/libengine/src/lib.rs

```diff
@@ -0,0 +1,318 @@
+// benchmarking state
+struct BenchState {
+    compilation_timer: *mut u8,
+    compilation_start: extern "C" fn(*mut u8),
+    compilation_end: extern "C" fn(*mut u8),
+    instantiation_timer: *mut u8,
+    instantiation_start: extern "C" fn(*mut u8),
+    instantiation_end: extern "C" fn(*mut u8),
+    execution_timer: *mut u8,
+    execution_start: extern "C" fn(*mut u8),
+    execution_end: extern "C" fn(*mut u8),
+    working_dir: PathBuf,
+    stdout_file: File,
+    stderr_file: File,
+}
+
+static mut NATIVE_EXECUTION_TIMER: Option<*mut u8> = None;
+static mut NATIVE_EXECUTION_START: Option<extern "C" fn(*mut u8)> = None;
+static mut NATIVE_EXECUTION_END: Option<extern "C" fn(*mut u8)> = None;
+
+#[no_mangle]
+pub extern "C" fn bench_start() {
+    let timer = unsafe { NATIVE_EXECUTION_TIMER.unwrap() };
+    unsafe { NATIVE_EXECUTION_START.unwrap()(timer) };
+}
+
+#[no_mangle]
+pub extern "C" fn bench_end() {
+    let timer = unsafe { NATIVE_EXECUTION_TIMER.unwrap() };
+    unsafe { NATIVE_EXECUTION_END.unwrap()(timer) };
+}
+
+impl BenchState {
+    fn new(
+        compilation_timer: *mut u8,
+        compilation_start: extern "C" fn(*mut u8),
+        compilation_end: extern "C" fn(*mut u8),
+        instantiation_timer: *mut u8,
+        instantiation_start: extern "C" fn(*mut u8),
+        instantiation_end: extern "C" fn(*mut u8),
+        execution_timer: *mut u8,
+        execution_start: extern "C" fn(*mut u8),
+        execution_end: extern "C" fn(*mut u8),
+        working_dir: PathBuf,
+        stdout_file: File,
+        stderr_file: File,
+    ) -> Result<Self> {
+        Ok(Self {
+            compilation_timer,
+            compilation_start,
+            compilation_end,
+            instantiation_timer,
+            instantiation_start,
+            instantiation_end,
+            execution_timer,
+            execution_start,
+            execution_end,
+            working_dir,
+            stdout_file,
+            stderr_file,
+        })
+    }
+
+    fn compile(&mut self, _bytes: &[u8]) -> Result<()> {
+        (self.compilation_start)(self.compilation_timer);
+        (self.compilation_end)(self.compilation_timer);
+        Ok(())
+    }
+
+    fn instantiate(&mut self) -> Result<()> {
+        (self.instantiation_start)(self.instantiation_timer);
+        (self.instantiation_end)(self.instantiation_timer);
+        Ok(())
+    }
+
+    fn execute(&mut self) -> Result<()> {
+        // Initialize call back timers
+        unsafe {
+            NATIVE_EXECUTION_TIMER = Some(self.execution_timer);
+            NATIVE_EXECUTION_START = Some(self.execution_start);
+            NATIVE_EXECUTION_END = Some(self.execution_end);
+        }
+
+        // Setup pipe for capturing stdout
+        let pipe_stdout = pipe()?;
+        let original_stdout = dup(1)?;
+        dup2(pipe_stdout.1, 1)?;
+        close(pipe_stdout.1)?;
+
+        // Load the library
+        let native_lib = self.working_dir.join("./benchmark.so");
```

Comment:
I feel like this path should have been a part of `wasm_bench_compile`; unfortunately that call gives us the bytes of the `.so`, not the path, which is what you need here. I wonder if we should change `wasm_bench_compile` to accept a path instead?

Reply:
Yes, I experimented with the path for loading the so file. Ultimately I did modify BenchState to include the path base tough the name of the so has to be benchmark.so. I don't know if wasm_bench_compile would be the place  but I do think including the basename of the so (or wasmfile) in wasm_bench_create would allow more flexibility. I didn't want to disrupt or potentially break the wasmtime support (just yet :-)) so I didn't worry about that for now.